lib/mfcdm/mfcdm/MediaFoundationCdmFactory.cpp

```diff
@@ -0,0 +1,213 @@
+    MultiByteToWideChar(CP_UTF8, 0, str.c_str(), (int)str.length(), &wstr[0], count);
+    return wstr;
+}
+
+MediaFoundationCdmFactory::MediaFoundationCdmFactory(std::string key_system)
+    : m_keySystem(std::move(key_system))
+{
+}
+
+void MediaFoundationCdmFactory::Initialize()
+{
+    const winrt::com_ptr<IMFMediaEngineClassFactory4> class_factory = winrt::create_instance<IMFMediaEngineClassFactory4>(
+      CLSID_MFMediaEngineClassFactory, CLSCTX_INPROC_SERVER);
+    const std::wstring key_system_str = ConvertUtf8ToWide(m_keySystem);
+
+    THROW_IF_FAILED(class_factory->CreateContentDecryptionModuleFactory(
+        key_system_str.c_str(), IID_PPV_ARGS(&m_cdmFactory)));
+}
+
+bool MediaFoundationCdmFactory::IsTypeSupported(const std::string &key_system) const {
+    return m_cdmFactory->IsTypeSupported(ConvertUtf8ToWide(key_system).c_str(), nullptr);
+}
+
+// Returns a property store similar to EME MediaKeySystemMediaCapability.
+void CreateVideoCapability(const media::CdmConfig& cdm_config,
+                              winrt::com_ptr<IPropertyStore>& video_capability)
+{
+    winrt::com_ptr<IPropertyStore> temp_video_capability;
+    THROW_IF_FAILED(
+            PSCreateMemoryPropertyStore(IID_PPV_ARGS(&temp_video_capability)));
+
+    ScopedPropVariant robustness;
+    if (cdm_config.use_hw_secure_codecs) {
+        PROPVARIANT* ptr = robustness.Receive();
+        ptr->vt = VT_BSTR;
+        ptr->bstrVal = SysAllocString(L"HW_SECURE_ALL");
+        THROW_IF_FAILED(
+                temp_video_capability->SetValue(MF_EME_ROBUSTNESS, robustness.get()));
+    }
+    video_capability = temp_video_capability;
+}
+
+/*!
+ * \link https://github.com/chromium/chromium/blob/ea198b54e3f6b0cfdd6bacbb01c2307fd1797b63/media/cdm/win/media_foundation_cdm_util.cc#L68
+ * \link https://github.com/microsoft/media-foundation/blob/969f38b9fff9892f5d75bc353c72d213da807739/samples/MediaEngineEMEUWPSample/src/media/eme/MediaKeySystemConfiguration.cpp#L74
+ */
+void BuildCdmAccessConfigurations(const media::CdmConfig& cdm_config,
+                                  winrt::com_ptr<IPropertyStore>& properties)
+{
+    winrt::com_ptr<IPropertyStore> temp_configurations;
+    THROW_IF_FAILED(PSCreateMemoryPropertyStore(IID_PPV_ARGS(&temp_configurations)));
+
+    PROPVARIANT* receive;
+
+    // Init data types
+    ScopedPropVariant init_types;
+
+    std::vector<const wchar_t*> init_types_array = {L"cenc"};
+
+    auto initDataTypeArray = static_cast<BSTR *>(CoTaskMemAlloc(sizeof(BSTR) * init_types_array.size()));
+    for(size_t i = 0; i < init_types_array.size(); i++)
+    {
+        initDataTypeArray[i] = SysAllocString(init_types_array[i]);
+    }
+    receive = init_types.Receive();
+    receive->vt = VT_VECTOR | VT_BSTR;
+    receive->cabstr.cElems = static_cast<ULONG>(init_types_array.size());
+    receive->cabstr.pElems = initDataTypeArray;
+    THROW_IF_FAILED(temp_configurations->SetValue(MF_EME_INITDATATYPES, init_types.get()));
```

Comment:
now i understand that you followed MS examples code, also to speed up writing and its can be okayish for a first test, but, use macro (like THROW_IF_FAILED)+exceptions are a bad code design, raise exceptions are discouraged, where they should be used where the software functionality is compromised/critical situations only

somewhat related to this there is also `std::cout` uses, that must be removed,
this should be avoided within kodi platform which has its own log system

Here there is some thinking to be done to implement better code, some tips:
- if you need to print to the log, you should provide your own library log interface, you can take ispiration from the `mpegts` library (under lib/) see debug.cpp, maybe you can even simplify that code, note that old code make use of defines like DEMUX_DBG_DEBUG that can be easily replaced with an enum. On src/TSReader.cpp you can see how mpegts debug interface is connected to addon log methods.
- Replace THROW_IF_FAILED will a bit more complex, methods that use that one, should return a status, more likely a simple "bool" should be enough to allow parent caller to act as needed. ofc FAILED/SUCCEEDED windows defines to check methods state are still needed because proprietary, but you will use within a condition, and only if needed, print something to log if fail, if helpful in solving a problem

at the end `MFException.h` and `utils/macro.h` should be deleted